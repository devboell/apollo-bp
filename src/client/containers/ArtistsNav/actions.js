/* eslint-disable import/prefer-default-export */
export const setSelectedArtistIndex = index => ({
  type: 'SET_SELECTED_ARTSIST_INDEX',
  index,
})
