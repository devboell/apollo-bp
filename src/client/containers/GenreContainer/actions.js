import types from './actionTypes'

export const setSelectedGenres = (name, genres) => ({
  type: types.SET_SELECTED_GENRES,
  name,
  genres,
})

export const selectGenre = (name, genre) => ({
  type: types.SELECT_GENRE,
  genre,
  name,
})

export const deselectGenre = (name, genre) => ({
  type: types.DESELECT_GENRE,
  genre,
  name,
})
