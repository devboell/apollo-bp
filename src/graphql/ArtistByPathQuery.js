import gql from 'graphql-tag'

export default gql`
  query ArtistByPathQuery($path: String) {
    artistByPath(path: $path) {
      path
      paintings {
        title
        name
        genre
      }
    }
  }
`
