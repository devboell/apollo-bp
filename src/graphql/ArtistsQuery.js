import gql from 'graphql-tag'

export default gql`
  query ArtistsQuery {
    artists {
      name
      path
    }
  }
`
