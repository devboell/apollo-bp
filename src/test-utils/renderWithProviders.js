import React from 'react'

import { render } from 'react-testing-library'
import { compose } from 'ramda'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history' /* eslint-disable-line */
import { MockedProvider } from 'react-apollo/test-utils'
import { Provider } from 'react-redux'


const withRouter = ([node, config]) => {
  const history = createMemoryHistory({ initialEntries: [config.route] })
  return [
    <Router history={history}>{node}</Router>,
    { ...config, history },
  ]
}

const withApollo = ([node, config]) => [
  <MockedProvider
    removeTypename
    mocks={config.mocks}
  >
    {node}
  </MockedProvider>,
  config,
]

const withRedux = ([node, config]) => [
  <Provider store={config.store}>{node}</Provider>,
  config,
]

const withProvider = key => fn => ([node, config]) => (
  !config[key]
    ? [node, config]
    : fn([node, config])
)

const withProviders = compose(
  withProvider('route')(withRouter),
  withProvider('mocks')(withApollo),
  withProvider('store')(withRedux),
)

const renderWithProviders = (node, config) => {
  const [WithProviders, { history }] = withProviders([node, config])
  return {
    ...render(WithProviders),
    history,
  }
}


// re-export everything
export * from 'react-testing-library'

// override render method
export { renderWithProviders }
