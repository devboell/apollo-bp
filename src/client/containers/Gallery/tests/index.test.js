import React from 'react'
import { MemoryRouter, withRouter } from 'react-router-dom'
import { MockedProvider } from 'react-apollo/lib/test-utils'
import configureStore from 'redux-mock-store'
import { print } from 'graphql'

import artistData from 'data/test/artistByPathQuery.json'
import * as actions from '../actions'

import GalleryWithData, {
  Gallery,
  ARTIST_BY_PATH_QUERY,
  withArtist,
  mapDispatchToProps,
} from '../index'


const middlewares = []
const mockStore = configureStore(middlewares)

describe('Gallery component', () => {
  const onSetSelectedGenresSpy = jest.fn()
  const otherProps = {
    selectedGenres: ['city', 'landscape'],
    onSetSelectedGenres: onSetSelectedGenresSpy,
    onSelectGenre: jest.fn(),
    onDeselectGenre: jest.fn(),
  }

  beforeEach(() => {
    onSetSelectedGenresSpy.mockClear()
  })

  it('with data snapshot', () => {
    const wrapper = shallow(
      <Gallery
        artist={artistData}
        loading={false}
        {...otherProps}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('without data snapshot', () => {
    const wrapper = shallow(
      <Gallery
        artist={{}}
        loading
        {...otherProps}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('calls onSetSelectedGenres for new aritst prop', () => {
    const wrapper = shallow(
      <Gallery
        artist={{}}
        loading
        {...otherProps}
      />,
    )
    wrapper.setProps({ artist: artistData })
    expect(onSetSelectedGenresSpy).toHaveBeenCalledTimes(1)
  })

  it('does not call onSetSelectedGenres for same aritst prop', () => {
    const wrapper = shallow(
      <Gallery
        artist={artistData}
        loading={false}
        {...otherProps}
      />,
    )
    wrapper.setProps({ artist: artistData })
    expect(onSetSelectedGenresSpy).toHaveBeenCalledTimes(0)
  })
})

describe('Gallery ARTIST_BY_PATH_QUERY', () => {
  it('has a known query shape', () => {
    expect(print(ARTIST_BY_PATH_QUERY)).toMatchSnapshot()
  })
})


/* eslint-disable react/prop-types */
describe('Gallery withArtist', () => {
  it('renders with loading first', () => {
    class Container extends React.Component {
      componentWillMount() {
        expect(this.props.loading).toBe(true)
        expect(this.props.artist).toBeFalsy()
      }
      render() {
        return null
      }
    }
    const ContainerWithData = withRouter(withArtist(Container))
    render(
      <MockedProvider
        mocks={[
          {
            request: { ARTIST_BY_PATH_QUERY },
            result: { data: artistData } },
        ]}
      >
        <MemoryRouter initialEntries={['/heroshige']}>
          <ContainerWithData />
        </MemoryRouter>
      </MockedProvider>,
    )
  })

  /* eslint-disable react/no-multi-comp */
  it('renders data without crashing', () => {
    class Container extends React.Component {
      componentWillReceiveProps(nextProps) {
        expect(nextProps.loading).toBe(false)
        expect(nextProps.artists).toEqual(artistData)
      }
      render() {
        return null
      }
    }
    const ContainerWithData = withRouter(withArtist(Container))
    render(
      <MockedProvider
        mocks={[
          {
            request: { ARTIST_BY_PATH_QUERY },
            result: { data: artistData } },
        ]}
      >
        <MemoryRouter initialEntries={['/heroshige']}>
          <ContainerWithData />
        </MemoryRouter>
      </MockedProvider>,
    )
  })
})

describe('Gallery, enhanced component', () => {
  it('renders without crashing', () => {
    const wrapper = render(
      <MockedProvider
        mocks={[
          { request: { ARTIST_BY_PATH_QUERY }, result: { data: artistData } },
        ]}
        store={mockStore({ gallery: { selectedGenres: [] } })}
      >
        <MemoryRouter initialEntries={['/heroshige']}>
          <GalleryWithData />
        </MemoryRouter>
      </MockedProvider>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('mapDispatchToProps', () => {
  const dispatch = jest.fn()
  const {
    onSetSelectedGenres,
    onSelectGenre,
    onDeselectGenre,
  } = mapDispatchToProps(dispatch)

  it('maps onSetSelectedGenres to setSelectedGenres', () => {
    const genres = ['city', 'landscape']
    const expected = actions.setSelectedGenres(genres)
    onSetSelectedGenres(genres)
    expect(dispatch).toHaveBeenCalledWith(expected)
  })

  it('maps onSelectGenre to selectGenre', () => {
    const genre = 'city'
    const expected = actions.selectGenre(genre)
    onSelectGenre(genre)
    expect(dispatch).toHaveBeenCalledWith(expected)
  })

  it('maps onDeselectGenre to deselectGenre', () => {
    const genre = 'city'
    const expected = actions.deselectGenre(genre)
    onDeselectGenre(genre)
    expect(dispatch).toHaveBeenCalledWith(expected)
  })
})
