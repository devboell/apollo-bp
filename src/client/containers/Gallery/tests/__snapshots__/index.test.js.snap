// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery ARTIST_BY_PATH_QUERY has a known query shape 1`] = `
"query ArtistByPathQuery($path: String) {
  artistByPath(path: $path) {
    path
    paintings {
      title
      name
      genre
    }
  }
}
"
`;

exports[`Gallery component with data snapshot 1`] = `
<styled.div>
  <GenreChooser
    genres={
      Array [
        "city",
        "landscape",
      ]
    }
    onDeselectGenre={[Function]}
    onSelectGenre={[Function]}
    selectedGenres={
      Array [
        "city",
        "landscape",
      ]
    }
  />
  <Component
    artistPath="heroshige"
    paintings={
      Array [
        Object {
          "genre": "city",
          "name": "cotton_goods_shops_at_odemmacho.jpg",
          "title": "Cotton goods shops at odemmacho",
        },
        Object {
          "genre": "landscape",
          "name": "sudden_shower_over_shin-ohashi_bridge_and_atake.jpg",
          "title": "Sudden Shower over Shin-ÅŒhashi bridge and Atake",
        },
      ]
    }
  />
</styled.div>
`;

exports[`Gallery component without data snapshot 1`] = `
<div>
  loading...
</div>
`;

exports[`Gallery, enhanced component renders without crashing 1`] = `
<div>
  loading...
</div>
`;
